include/gzip/compress.hpp

```diff
@@ -57,7 +59,7 @@ std::string compress(const char* data,
         // There is no way we see that "increase" would not fit in an unsigned int,
         // hence we use static cast here to avoid -Wshorten-64-to-32 error
         deflate_s.avail_out = static_cast<unsigned int>(increase);
-        deflate_s.next_out = (Bytef*)(output.data() + length);
+        deflate_s.next_out = reinterpret_cast<Bytef*>((&output[0] + length));
```

Comment:
@springmeyer mind explaining why `reinterpret_cast` instead of static or another?

Reply:
Proximately because a `static_cast` would not compile:

```
/Users/dane/projects/gzip-hpp/include/gzip/compress.hpp:62:30: error: 
      static_cast from 'value_type *' (aka 'char *') to 'Bytef *'
      (aka 'unsigned char *') is not allowed
        deflate_s.next_out = static_cast<Bytef*>((&output[0] + length));
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Which means the previous `C` style cast was doing the equivalent of a `reinterpret_cast` under the hood. A `reinterpret` is needed because we are changing a `char *` to an `unsigned char *`. We are doing this because a `char *` is what is stored inside `std::string` per http://en.cppreference.com/w/cpp/string/basic_string. While zlib's `next_out` uses a `Bytef` pointer and a `Bytef` is type def'd to a `Byte` which is type def'd to an `unsigned char` at https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/zconf.h#L391.

So our options are either:

 - Make the types the same
 - Cast

In this case I'm casting since that is what the previous code did. Also because the only way I can think of making the types the same is to use a `std::vector<unsigned char>` instead of a `std::string` and I think `std::string` is ideal since that is what has been in use.

If you want to know the difference between `char *` and `unsigned char*` and why zlib uses `unsigned` I'm not sure - I'd need to tag in @joto for an opinion.